marshal/admin.go

```diff
@@ -0,0 +1,358 @@
+	defer close(fail)
+	var wg sync.WaitGroup
+	for _, claim := range a.claims {
+		wg.Add(1)
+
+		releaseOffset := claim.currentOffset
+		// Update our step counter. Releasing with a new offset will Heartbeat then ReleasePartition,
+		// releasing without resetting the offset will just ReleasePartition
+		atomic.AddInt32(a.steps, 1)
+		if newOffset {
+			releaseOffset = claim.newOffset
+			atomic.AddInt32(a.steps, 1)
+		}
+
+		go func(t string, p int, offset int64, newOffset bool) {
+			if ok := a.release(t, p, offset, newOffset); !ok {
+				fail <- true
+			}
+			wg.Done()
+		}(claim.topic, claim.partID, releaseOffset, newOffset)
+	}
+
+	// Wait on all workers to reset their respective Kafka offset.
+	done := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+
+	select {
+	case <-fail:
+		return fmt.Errorf("Admin failed to reset Kafka offset!")
+	case <-done:
+		return nil
+	}
+}
+
+// heartbeatLoop hearbeats as if we had a claim to this partition and were simply
+// not reading past where the previous owner had left off.
+func (a *consumerGroupAdmin) heartbeatLoop(
+	topic string, partID int, lastOffset int64, stopChan chan struct{}) {
+
+	for {
+		select {
+		// Stop claimHealth either when all topic, partitions have been successfully claimed,
+		// or the Admin has failed to do so and needs to abort.
+		case <-stopChan:
+			return
+		default:
+			// If we fail to heartbeat, record this in claimHealth.
+			// The Admin will take care of cleaning up other claims.
+			if err := a.marshaler.Heartbeat(topic, partID, lastOffset); err != nil {
+				log.Errorf("[%s:%d] Admin failed to heartbeat. It is now unhealthy "+
+					"and will not reset offsets.", topic, partID)
+				atomic.StoreInt32(a.claimHealth, 0)
+				return
+			}
+			atomic.AddInt32(a.steps, 1)
+			time.Sleep(<-a.marshaler.cluster.jitters)
+		}
+	}
+}
+
+// claimAndHeartbeat attempts to claim a partition released by a paused consumer.
+// It heartbeats the previous offset.
+func (a *consumerGroupAdmin) claimAndHeartbeat(topic string,
+	partID int, newOffset int64, stopHeartbeat chan struct{}) bool {
+
+	// Get current offsets, which we will try to keep claimHealth.
+	partitionClaim := a.marshaler.GetLastPartitionClaim(topic, partID)
+
+	// Next, try to claim the partition.
+	if !a.marshaler.ClaimPartition(topic, partID) {
+		log.Errorf("[%s:%d] Admin couldn't claim partition to set Kafka offset",
+			topic, partID)
+		return false
+	}
+	atomic.AddInt32(a.steps, 1)
+
+	// Continuously heartbeat the last offsets.
+	a.addClaimAttempt(topic, partID, partitionClaim.CurrentOffset, newOffset)
+	go a.heartbeatLoop(topic, partID, partitionClaim.CurrentOffset, stopHeartbeat)
+	return true
+}
+
+// constructReleaseGroupMessage returns a ReleaseGroup message to write to the Marshal topic.
+func (a *consumerGroupAdmin) constructReleaseGroupMessage() *msgReleaseGroup {
+	now := time.Now()
+	base := &msgBase{
+		Time:       int(now.Unix()),
+		InstanceID: newInstanceID(),
```

Comment:
Hmmm this means every ReleaseGroup message has a new instance ID which seems odd. Why are you doing this?